From 4200f7b0f83936fa5e498c147cc2527841efa1f4 Mon Sep 17 00:00:00 2001
From: Juri Lelli <juri.lelli@redhat.com>
Date: Tue, 16 Feb 2021 16:11:38 +0100
Subject: [PATCH] utils: Add HRTICK_DL support

[ commit upstream 8d4390201a2234cb095c5cb8bcaa22c99a0ccb09 ]

Recent kernels have added distinction between HRTICK (for NORMAL tasks)
and HRTICK_DL (for DEADLINE tasks).

Use the latter (if available) when needed for boosting tasks using
DEADLINE policy.

Signed-off-by: Juri Lelli <juri.lelli@redhat.com>
Signed-off-by: Sharan Turlapati <sturlapati@vmware.com>
---
 src/utils.c | 32 ++++++++++++++++++++++++++------
 1 file changed, 26 insertions(+), 6 deletions(-)

diff --git a/src/utils.c b/src/utils.c
index cecefe0..5af58e2 100644
--- a/src/utils.c
+++ b/src/utils.c
@@ -385,6 +385,7 @@ int setup_hr_tick(void)
 	int ret;
 	int len;
 	int fd;
+	int hrtick_dl = 0;
 
 	if (set)
 		return 1;
@@ -419,19 +420,38 @@ int setup_hr_tick(void)
 
 	ret = 1;
 
-	p = strstr(buf, "HRTICK");
-	if (p + 3 >= buf) {
+	p = strstr(buf, "HRTICK_DL");
+	if (p && p - 3 >= buf) {
+		hrtick_dl = 1;
 		p -= 3;
-		if (strncmp(p, "NO_HRTICK", 9) == 0) {
-			log_msg("dl_runtime is shorter than 1ms, setting HRTICK\n");
-			ret = write(fd, "HRTICK", 6);
-			if (ret != 6)
+		if (strncmp(p, "NO_HRTICK_DL", 12) == 0) {
+			log_msg("dl_runtime is shorter than 1ms, setting HRTICK_DL\n");
+			ret = write(fd, "HRTICK_DL", 9);
+			if (ret != 9)
 				ret = 0;
 			else
 				ret = 1;
 		}
 	}
 
+	/*
+	 * Backward compatibility on kernels that only have HRTICK.
+	 */
+	if (!hrtick_dl) {
+		p = strstr(buf, "HRTICK");
+		if (p && p - 3 >= buf) {
+			p -= 3;
+			if (strncmp(p, "NO_HRTICK", 9) == 0) {
+				log_msg("dl_runtime is shorter than 1ms, setting HRTICK\n");
+				ret = write(fd, "HRTICK", 6);
+				if (ret != 6)
+					ret = 0;
+				else
+					ret = 1;
+			}
+		}
+	}
+
 	close(fd);
 	return ret;
 }
-- 
2.19.0

